src/test/kotlin/ch/sourcemotion/vertx/kinesis/consumer/orchestra/consumer/RecordFetcherTest.kt

```diff
@@ -74,9 +77,106 @@ internal class RecordFetcherTest : AbstractVertxTest() {
+    }
+
+    @Test
+    internal fun getNextRecords_should_be_delayed(testContext: VertxTestContext) = asyncTest(testContext) {
+        val responseMock: GetRecordsResponse = mock {
+            on { nextShardIterator() } doReturn ITERATOR
+        }
+
+        val kinesisClientMock = mock<KinesisAsyncClient> {
+            on { getRecords(any<Consumer<GetRecordsRequest.Builder>>()) } doReturn CompletableFuture.completedFuture(
+                responseMock
+            )
+        }
+
+        val delay = 100L
 
-        val response = sut.fetchNextRecords(queryPosition)
-        response.nextShardIterator().shouldBe(nextIterator)
+        val sut = createFetcher(kinesisClientMock, delay)
+
+        val start = System.currentTimeMillis()
+
+        sut.fetchNextRecords(queryPosition)
+        sut.getNextRecords()
+
+        val end = System.currentTimeMillis()
+        val diff = end - start
+        diff.shouldBeGreaterThanOrEqual(delay)
     }
+
+    @Test
+    internal fun exception_case_restart_fetching(testContext: VertxTestContext) = asyncTest(testContext) {
+        val responseMock: GetRecordsResponse = mock {
+            on { nextShardIterator() } doReturn ITERATOR
+        }
+
+        val previousShardIterator = "previous-shard-iterator"
+        val previousShardIteratorResponseMock: GetRecordsResponse = mock {
+            on { nextShardIterator() } doReturn previousShardIterator
+        }
+
+        val kinesisClientMock = mock<KinesisAsyncClient> {
+            on { getRecords(any<Consumer<GetRecordsRequest.Builder>>()) } doAnswer {
+                val builder = GetRecordsRequest.builder()
+                (it.arguments.first() as Consumer<GetRecordsRequest.Builder>).accept(builder)
+                val req = builder.build()
+                val requestedShardIterator = req.shardIterator()
+                if (requestedShardIterator == ITERATOR) {
+                    CompletableFuture.completedFuture(responseMock)
+                } else {
+                    CompletableFuture.completedFuture(previousShardIteratorResponseMock)
+                }
+            }
+        }
+
+        val sut = createFetcher(kinesisClientMock)
+
+        sut.fetchNextRecords(queryPosition)
+
+        val previousPosition = QueryPosition(
+            ShardIterator(previousShardIterator),
+            SequenceNumber("sequence-number", SequenceNumberIteratorPosition.AFTER)
+        )
+
+        sut.restartFetching(previousPosition)
+
+        val response = sut.getNextRecords()
+        response.nextShardIterator().shouldBe(previousShardIterator)
+        verify(kinesisClientMock, times(2)).getRecords(any<Consumer<GetRecordsRequest.Builder>>())
+    }
+
+    @Test
+    internal fun wrong_fetching_access_order(testContext: VertxTestContext) = asyncTest(testContext) {
+        val kinesisClientMock = mock<KinesisAsyncClient> { }
+        val sut = createFetcher(kinesisClientMock)
+        assertFailsWith<RecordFetcher.WrongFetchingOrderException> {
```

Comment:
Perhaps some default value is in effect? It is totally not clear why such sequence of actions should lead to an exception